src/NuGet.Core/NuGet.ProjectModel/DependencyGraphSpec.cs

```diff
@@ -43,34 +54,64 @@ public DependencyGraphSpec(bool isReadOnly)
             _isReadOnly = isReadOnly;
         }
 
+        private bool _refreshRestoreList = false;
+        private IReadOnlyList<string>? _restoreList = null;
+
         /// <summary>
         /// Projects to restore.
         /// </summary>
-        public IReadOnlyList<string> Restore => _restore.ToList();
+        public IReadOnlyList<string> Restore
+        {
+            get
+            {
+                if (_restoreList == null || _refreshRestoreList)
+                {
+                    _restoreList = _restore.ToList();
```

Comment:
Does this property get called once per restore, or can it get called multiple times?

Specifically what I'm thinking about is if this method is going to get called multiple times concurrently when `_restoreList == null || _refreshRestoreList`.

This method is idempotent, so this implementation won't cause correctness problems. But `.ToList()` will be called my multiple threads concurrently, causing the same list to be allocated multiple times. If the backing store doesn't implement `ICollection`, then `.ToList()` is going to cause multiple array allocations for the list's backing store, so multiple threads doing so concurrently will have a particularly bad allocation effect.

Basicly, is it worth wrapping the inside of the `if` statement in a `lock`, to reduce memory allocations?

Reply:
It gets called multiple times in restore, but it's the population is not concurrent (each project has it's own dg spec, so each project enumerates early). 

I haven't seen it called concurrently, it's just called a bunch of times during restore cause we didn't know any better.